### PR TITLE
Introduced DockerClassRule

### PR DESCRIPTION
**Pull request by jglick**
Got tired of hitting 3m test timeouts inside slow `docker-build` processes.

@reviewbybees

### CODE REVIEW

**Comment by olivergondza** [comment]
Is this all about getting the container build out of timeout constrain? Do we want this to be done everytime (deprecating the `@Rule` usage)?

**Comment by jglick** [comment]
> Is this all about getting the container build out of timeout constrain?

Yes.

> Do we want this to be done everytime (deprecating the `@Rule` usage)?

I consider deprecating `DockerRule` but decided to leave it for now, since there may be tests which were not subject to timeout anyway and were happy with the original behavior. I did switch the documentation to suggest `DockerClassRule`. No strong opinion.

BTW the class rule has a minor disadvantage in that it does not stop the containers until the end of the whole suite—apparently JUnit offers no API to allow a rule to take actions both before the suite starts, and after each test ends: you would have to create two rules and have the user register both.

Another approach I played with was adding an `eager` flag to `DockerRule` (making it build the image before starting the base statement), and then having the user use `RuleChain` to make sure the `DockerRule` comes “outside” the `JenkinsRule`. This had a minor disadvantage in that it would run `docker-build` for each test case. (All but the first being guaranteed to use the Docker cache and so run pretty quickly.) More to the point, it was just harder to use and harder to convert existing tests to. By contrast, switching `@Rule public DockerRule` to `@ClassRule public static DockerClassRule` (and one or `get` calls to one `create` call) seemed minimally intrusive.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerClassRule.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    @Override
+    public Statement apply(final Statement base, Description description) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                image = delegate.build();
+                try {
+                    base.evaluate();
+                } finally {
+                    containers.forEach(T::close);
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerClassRule.java]: Neat, I didn't know it was possible to use a type parameter in a method reference like that.

Comment (jglick) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerClassRule.java]: Neither did I until I tried it and it compiled. :-)

**Pull request merged**